paymentsheet/src/test/java/com/stripe/android/paymentsheet/ExternalPaymentMethodProxyActivityTest.kt

```diff
@@ -0,0 +1,142 @@
+        )
+
+        idleLooper()
+
+        assertThat(confirmedType).isEqualTo(expectedExternalPaymentMethodType)
+        assertThat(confirmedBillingDetails).isEqualTo(expectedBillingDetails)
+        runCatching {
+            scenario?.result
+        }.onSuccess {
+            fail(message = "Activity has finished, but it should not have finished yet!")
+        }
+    }
+
+    @Test
+    fun `ExternalPaymentMethodResultHandler finishes EPM activity with correct result on completed`() {
+        val scenario = ActivityScenario.launchActivityForResult<ExternalPaymentMethodProxyActivity>(
+            ExternalPaymentMethodResultHandler.createResultIntent(context, ExternalPaymentMethodResult.completed())
+        )
+
+        assertThat(scenario.result.resultCode).isEqualTo(ExternalPaymentMethodResult.Completed.RESULT_CODE)
+    }
+
+    @Test
+    fun `ExternalPaymentMethodResultHandler finishes EPM activity with correct result on canceled`() {
+        val scenario = ActivityScenario.launchActivityForResult<ExternalPaymentMethodProxyActivity>(
+            ExternalPaymentMethodResultHandler.createResultIntent(context, ExternalPaymentMethodResult.canceled())
+        )
+
+        assertThat(scenario.result.resultCode).isEqualTo(ExternalPaymentMethodResult.Canceled.RESULT_CODE)
+    }
+
+    @Test
+    fun `ExternalPaymentMethodResultHandler finishes EPM activity with correct result on failed`() {
+        val expectedDisplayMessage = "error!"
+        val scenario = ActivityScenario.launchActivityForResult<ExternalPaymentMethodProxyActivity>(
+            ExternalPaymentMethodResultHandler.createResultIntent(
+                context,
+                ExternalPaymentMethodResult.failed(displayMessage = expectedDisplayMessage)
+            )
+        )
+
+        assertThat(scenario.result.resultCode).isEqualTo(ExternalPaymentMethodResult.Failed.RESULT_CODE)
+        assertThat(
+            scenario.result.resultData.getStringExtra(ExternalPaymentMethodResult.Failed.DISPLAY_MESSAGE_EXTRA)
+        ).isEqualTo(expectedDisplayMessage)
+    }
+
+    @Test
+    fun `Start ExternalPaymentMethodProxyActivity with unexpected intent and it immediately finishes`() {
+        val scenario = ActivityScenario.launchActivityForResult<ExternalPaymentMethodProxyActivity>(
+            Intent().setClass(context, ExternalPaymentMethodProxyActivity::class.java)
+        )
+
+        assertThat(scenario.result).isNotNull()
+    }
+
+    class DefaultExternalPaymentMethodConfirmHandler(
+        private val setConfirmedType: (String?) -> Unit,
+        private val setConfirmedBillingDetails: (PaymentMethod.BillingDetails?) -> Unit,
+    ) : ExternalPaymentMethodConfirmHandler {
+        override fun confirmExternalPaymentMethod(
+            externalPaymentMethodType: String,
+            billingDetails: PaymentMethod.BillingDetails
+        ) {
+            setConfirmedType(externalPaymentMethodType)
+            setConfirmedBillingDetails(billingDetails)
```

Comment:
Ideally, this would be a more realistic confirm handler, in that it would call `onExternalPaymentMethodResult` with some result. Then we would verify that the activity had finished with the correct result. Unfortunately, it doesn't look like ActivityScenario provides an API to send a new Intent to an already created activity, so I can't test the behavior of sending the second intent with the external payment method result extra. 

Will need full integration tests for that. These tests provide good coverage of ExternalPaymentMethodActivityTest responding correctly to the two types of Intent that it expects to receive

Reply:
There's some espresso APIs that might help: https://developer.android.com/training/testing/espresso/intents#stubbing